app/controllers/propertyLinking/UploadController.scala

```diff
@@ -0,0 +1,117 @@
+      case EvidenceChoices.OTHER      =>
+        Ok(uploadEvidence(session.submissionId, List.empty, session.uploadEvidenceData.attachments.getOrElse(Map.empty), session.evidenceType.map(x => form.fill(x)).getOrElse(form)))
+      case _                          =>
+        BadRequest(errorHandler.badRequestTemplate)
+    }
+  }
+
+  def initiate(evidence: UploadEvidenceChoiceParameters): Action[JsValue] = authenticatedAction.async(parse.json) { implicit request =>
+    withJsonBody[InitiateAttachmentRequest] { attachmentRequest =>
+      (for {
+        initiateAttachmentResult <- businessRatesAttachmentsServices.initiateAttachmentUpload(InitiateAttachmentPayload(attachmentRequest, routes.UploadController.show(evidence).absoluteURL()))
+      } yield Ok(Json.toJson(initiateAttachmentResult)))
+        .recover {
+          case _: FileAttachmentFailed  =>
+            BadRequest(Json.toJson(Messages("error.businessRatesAttachment.does.not.support.file.types")))
+          case ex: Exception            =>
+            Logger.warn("FileAttachmentFailed Exception:", ex)
+            InternalServerError("500 INTERNAL_SERVER_ERROR")
+        }
+    }
+  }
+
+  def continue(evidence: UploadEvidenceChoiceParameters): Action[AnyContent] = authenticatedAction.andThen(withLinkingSession).async { implicit request =>
+    def upload(uploadedData: UploadEvidenceData)(implicit request: LinkingSessionRequest[_]): Option[Future[Result]] = {
+      PartialFunction.condOpt(request.ses.uploadEvidenceData.attachments) {
+        case Some(fileData) if fileData.nonEmpty =>
+          businessRatesAttachmentsServices.persistSessionData(request.ses, uploadedData).map(x => Redirect(routes.Declaration.show().url))
+      }
+    }
+
+    val session = request.ses
+    evidence.choice match {
+      case EvidenceChoices.RATES_BILL =>
+        upload(session.uploadEvidenceData.copy(linkBasis = RatesBillFlag, fileInfo = session.uploadEvidenceData.fileInfo.map(_.copy(evidenceType = RatesBillType))))
+          .getOrElse(Future.successful(BadRequest(uploadRatesBill(request.ses.submissionId, List("error.businessRatesAttachment.file.not.selected"), Map()))))
+      case EvidenceChoices.OTHER =>
+        form.bindFromRequest().fold(
+          _ =>
+            Future.successful(BadRequest(uploadEvidence(request.ses.submissionId, List("error.businessRatesAttachment.evidence.not.selected"), request.ses.uploadEvidenceData.attachments.getOrElse(Map()), form))),
+          formData => {
+            val updatedSession = session.copy(evidenceType = EvidenceType.fromName(formData.name))
+            val sessionUploadData: UploadEvidenceData = updatedSession.uploadEvidenceData
+              .copy(
+                linkBasis = OtherEvidenceFlag,
+                fileInfo = updatedSession.uploadEvidenceData.fileInfo.map(_.copy(evidenceType = EvidenceType.fromName(formData.name).get))) //.get is dangerous,
```

Comment:
agree, `.get` is asking for trouble. basically, if you can't bind a valid EvidenceType from request, then user should go to BadRequest page. at this point you should have a valid EvidenceType. Resolving it from String here leads to having to do .get which can explode